staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/controller.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package responsewriter
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+)
+
+// the user-provided decorator that contains the http.ResponseController object
+// must provide the UnwrapResponseController method so it can be accessed
+// by any http Handler in the chain.
+type ResponseControllerWrapper interface {
+	UnwrapResponseController() *ResponseController
+}
+
+// GetResponseController goes through the chain of wrapped http.ResponseWriter
+// objects and returns the original http.ResponseController object that was
+// added for a given request.
+//
+// any HTTP handler in the chain should be able to get the underlying
+// http.ResponseController object and manipulate Write and Read deadline
+// per request handler, as shown in the following example:
```

Comment:
naive question, 
do we want to do this?
to provide this to each handler or only use it in the timeout handler?

Reply:
Good question, I am not aware of any use case where we should allow each handler to manipulate Read/Write deadline. Also, letting each handler manipulate Read/Write deadline comes with more risks/complexity - we need to make sure they don't set a deadline that is larger than the overall request deadline.
In my opinion, we should not expose it to other handlers, not unless we have a clear need for it. I just wanted to demonstrate how it can be done if we needed to.

Reply:
yes, I agree, thanks!